### PR TITLE
Implement new Science Station protocol

### PR DESCRIPTION
**Pull request by evstraw**
Sorry for the delay - this PR introduces packet assembly functions and constants for the new science packets that we agreed on at the meeting yesterday. Notable changes are
- Addition of lazy susan position set packet
- Removal of science-specific telemetry packets; common telemetry packets should be used instead. I noticed that the current way telemetry codes are set up is there are some seemingly board-specific codes that are defined to mean something else in `CANScience.h`; I followed this convention in this PR, but wanted some comments on whether this is a good system or not.
- Reworking of continuous rotation servo set packet - power and direction are now encoded in a single signed value that ranges from -100 to 100.
- Documentation in the header file and wiki (https://github.com/huskyroboticsteam/HindsightCAN/wiki/Science-(Sensor)-Board-Packets). Unfortunately, GitHub doesn't let us include wiki changes as part of a PR, so I had to just go ahead and change that - please let me know if anything is inaccurate. I started putting version numbers at the top of pages I edited, so we can have some idea of whether it is up to date.

### CODE REVIEW

**Review by CaseyHackerMan — COMMENTED** [review]
Most continuous servo libraries I've seen have a range of 0-180 instead of the -100-100 we have here, with a value around 90 meaning no rotation. I think we should use this.

I also saw that we removed "AssembleScienceSensorPullPacket()". I assume this is because we are moving to periodic telemetry, and using a telemetry timing packet to start it. However, I don't see anything along the lines of "AssembleScienceTelemetryPacket()" that would be sent periodically.

**Comment by evstraw** [comment]
> Most continuous servo libraries I've seen have a range of 0-180 instead of the -100-100 we have here, with a value around 90 meaning no rotation. I think we should use this.

I feel like semantically, it makes more sense to use a signed integer to make the idea of speed/direction clearer; direction is represented clearly by the sign, speed is the absolute values, and zero intuitively corresponds to zero power. We could perhaps lower the range to 90, and internally, the firmware could shift values to be in the 0-180 range by adding 90 - what do you think?

> I also saw that we removed "AssembleScienceSensorPullPacket()". I assume this is because we are moving to periodic telemetry, and using a telemetry timing packet to start it. However, I don't see anything along the lines of "AssembleScienceTelemetryPacket()" that would be sent periodically.

The "science telemetry" packets were actually just standard telemetry packets, which are already implemented as common packets (actually, under the hood, the packets were identical and the `AssembleScienceSensorPullPacket()` function was just calling `AssembleTelemetryPullPacket()`), so I removed the science-specific versions in favor of just using the common versions.
For telemetry pull packets, `AssembleTelemetryPullPacket` (in `CANCommon.h`) should be used, with one of the `CAN_SCIENCE_SENSOR_...` constants as the telemetry code; for telemetry reports, `AssembleTelemetryReportPacket` (in `CANCommon.h`) should be used. I figured it was better to avoid duplicating packets we already had.

**Comment by abhaybd** [comment]
> > Most continuous servo libraries I've seen have a range of 0-180 instead of the -100-100 we have here, with a value around 90 meaning no rotation. I think we should use this.
> 
> I feel like semantically, it makes more sense to use a signed integer to make the idea of speed/direction clearer; direction is represented clearly by the sign, speed is the absolute values, and zero intuitively corresponds to zero power. We could perhaps lower the range to 90, and internally, the firmware could shift values to be in the 0-180 range by adding 90 - what do you think?

I agree with Evan, I think the sign better communicates the direction of rotation. The fact that CRs normally use 0-180 is just an implementation detail due to the fact that to the firmware, CR servos are indistinguishable from normal servos. However, since this packet is specific to CR servos, it makes sense to use a range that matches the intuition of positive = forward, negative = backwards, 0 = off.

File: CANScience.c [hunk 1]
```diff
@@ -1,56 +1,59 @@
...
+	return packet->data[2];
+}
+
+uint8_t GetScienceLazySusanPosFromPacket(CANPacket *packet) {
+    return packet->data[1];
+}
+
+uint8_t GetScienceServoIDFromPacket(CANPacket *packet){
+	return packet->data[1];
+}
```
Comment (abhaybd) [inline CANScience.c]: Can we make the parameter type `const CANPacket *`? No reason to not have a `const` here.

**Review by abhaybd — CHANGES_REQUESTED** [review]
Overall looks great, just some minor nitpicks about parameter types.

**Comment by evstraw** [comment]
Do we want to change the range to [-90, 90] to make it easier to translate to firmware? Not like we get (or need) a ton of precision with CR servos (and everything will likely be in the continuous interval [-1,1] on the Software side of things anyway since that's how we like to represent power values)

**Comment by abhaybd** [comment]
> Do we want to change the range to [-90, 90] to make it easier to translate to firmware? Not like we get (or need) a ton of precision with CR servos (and everything will likely be in the continuous interval [-1,1] on the Software side of things anyway since that's how we like to represent power values)

Sure, we could do that. Imo [-100,100] is a bit more intuitive than [-90,90], but as long as it's documented it should be fine. Especially if it makes the firmware implementation easier.

File: CANScience.c (hunk 1 above)
Comment (JohnDepaszthory) [reply CANScience.c]: This should honestly be done consistently throughout the entire HindsightCAN library. None of the other CAN* files use `const` in this case either

**Comment by JohnDepaszthory** [comment]
One concern that I always have about "unnecessarily" scaling values in applications like this it be rather inefficient on embedded platforms. For example, neither the PSoC nor AVR have hardware FPUs so either you need to use clever integer math or let the compiler use software floats - which is slower and may increase code size a fair bit because it will need to link the soft float library. In my code, I completely avoided floats to avoid needing to link the softfloat lib

**Comment by clowdur** [comment]
> Do we want to change the range to [-90, 90] to make it easier to translate to firmware? Not like we get (or need) a ton of precision with CR servos (and everything will likely be in the continuous interval [-1,1] on the Software side of things anyway since that's how we like to represent power values)

I think -100 to 100 makes more sense as a percentage of the max speed since angles are irrelevant to CR servos. Using -90 to 90 doesn't actually make much of a difference for firmware. As for efficiency, I think we would always have to scale things anyway since the servos have different PWM ranges, though I am probably misunderstanding. Also, sorry for closing and restoring a bunch of stuff, this is the first time I have used this I chose the wrong button to press like 4 times.

**Review by clowdur — APPROVED** [review]
LGTM!

**Comment by evstraw** [comment]
Sorry I took so long on this - I should have addressed all the comments. Let me know if there's anything else.

**Review by abhaybd — APPROVED** [review]
LGTM!

**Pull request merged**